modules/criteoBidAdapter.js

```diff
@@ -31,12 +32,108 @@ const PUBLISHER_TAG_URL_TEMPLATE = 'https://static.criteo.net/js/ld/publishertag
 const FAST_BID_PUBKEY_E = 65537;
 const FAST_BID_PUBKEY_N = 'ztQYwCE5BU7T9CDM5he6rKoabstXRmkzx54zFPZkWbK530dwtLBDeaWBMxHBUT55CYyboR/EZ4efghPi3CoNGfGWezpjko9P6p2EwGArtHEeS4slhu/SpSIFMjG6fdrpRoNuIAMhq1Z+Pr/+HOd1pThFKeGFr2/NhtAg+TXAzaU=';
 
+const SID_COOKIE_NAME = 'cto_sid';
+const IDCPY_COOKIE_NAME = 'cto_idcpy';
+const LWID_COOKIE_NAME = 'cto_lwid';
+const OPTOUT_COOKIE_NAME = 'cto_optout';
+const BUNDLE_COOKIE_NAME = 'cto_bundle';
+const GUID_RETENTION_TIME_HOUR = 24 * 30 * 13; // 13 months
+const OPTOUT_RETENTION_TIME_HOUR = 5 * 12 * 30 * 24; // 5 years
+
 /** @type {BidderSpec} */
 export const spec = {
   code: BIDDER_CODE,
   gvlid: GVLID,
   supportedMediaTypes: [BANNER, VIDEO, NATIVE],
 
+  getUserSyncs: function (syncOptions, _, gdprConsent, uspConsent) {
+    const fastBidVersion = config.getConfig('criteo.fastBidVersion');
+    if (canFastBid(fastBidVersion)) {
+      return [];
+    }
+
+    const refererInfo = getRefererInfo();
+    const origin = 'criteoPrebidAdapter';
+
+    if (syncOptions.iframeEnabled && hasPurpose1Consent(gdprConsent)) {
+      const queryParams = [];
+      queryParams.push(`origin=${origin}`);
+      queryParams.push(`topUrl=${refererInfo.domain}`);
+      if (gdprConsent) {
+        if (gdprConsent.gdprApplies) {
+          queryParams.push(`gdpr=${gdprConsent.gdprApplies == true ? 1 : 0}`);
+        }
+        if (gdprConsent.consentString) {
+          queryParams.push(`gdpr_consent=${gdprConsent.consentString}`);
+        }
+      }
+      if (uspConsent) {
+        queryParams.push(`us_privacy=${uspConsent}`);
+      }
+
+      const requestId = Math.random().toString();
+
+      const jsonHash = {
+        bundle: readFromAllStorages(BUNDLE_COOKIE_NAME),
+        cw: storage.cookiesAreEnabled(),
+        localWebId: readFromAllStorages(LWID_COOKIE_NAME),
+        lsw: storage.localStorageIsEnabled(),
+        optoutCookie: readFromAllStorages(OPTOUT_COOKIE_NAME),
+        origin: origin,
+        requestId: requestId,
+        secureIdCookie: readFromAllStorages(SID_COOKIE_NAME),
+        tld: refererInfo.domain,
+        topUrl: refererInfo.domain,
+        uid: readFromAllStorages(IDCPY_COOKIE_NAME),
+        version: '$prebid.version$'.replace(/\./g, '_'),
+      };
+
+      window.addEventListener('message', function (event) {
```

Comment:
this leaves a dangling listener each time it's run; imo it should detach itself when it gets the matching requestId back

Reply:
I'm also curious: how does the sync figure out which window to message? this is the first time I see this approach and I'm wondering if it needs support from core for the general case (where prebid might be in a window other than `top`).

Reply:
From the sync frame, we simply try to send the message to window.parent (which should be the one listening to the message as it's the one which dropped the frame)